tesseract/evm/src/gas_oracle.rs

```diff
@@ -295,29 +305,61 @@ fn get_cost_of_one_wei(eth_usd: U256) -> U256 {
 	eth_usd / eth_to_wei
 }
 
-/// Returns the L2 data cost for a given transaction data in usd
+/// Returns the L2 data cost for a given transaction data in USD.
+/// Implementation follows Optimism's Ecotone gas model:
+/// - Base implementation: https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/contracts/L2/GasPriceOracle.sol
+/// - Ecotone upgrade specs: https://github.com/ethereum-optimism/optimism/blob/develop/specs/protocol/fees.md
+/// 
+/// The total cost consists of:
+/// 1. L1 data fee: calculated using L1 base fee and the base fee scalar
+/// 2. Blob fee (Ecotone only): additional fee for blob data based on blob base fee and scalar
 pub async fn get_l2_data_cost(
-	rlp_tx: Bytes,
-	chain: StateMachine,
-	client: Arc<Provider<Http>>,
-	// Unit wei cost in 27 decimals
-	unit_wei_cost: U256,
+    rlp_tx: Bytes,
+    chain: StateMachine,
+    client: Arc<Provider<Http>>,
+    unit_wei_cost: U256,
 ) -> Result<Cost, anyhow::Error> {
-	let mut data_cost = U256::zero();
-	match chain {
-		StateMachine::Evm(inner_evm) => match inner_evm {
-			id if is_op_stack(id) => {
-				let ovm_gas_price_oracle = OVM_gasPriceOracle::new(H160(OP_GAS_ORACLE), client);
-				let data_cost_bytes = ovm_gas_price_oracle.get_l1_fee(rlp_tx).await?; // this is in wei
-				data_cost = data_cost_bytes * unit_wei_cost
-			},
-
-			_ => {},
-		},
-		_ => Err(anyhow!("Unknown chain: {chain:?}"))?,
-	}
-
-	Ok(convert_27_decimals_to_18_decimals(data_cost)?.into())
+    let mut data_cost = U256::zero();
+    match chain {
+        StateMachine::Evm(inner_evm) => match inner_evm {
+            id if is_op_stack(id) => {
+                let ovm_gas_price_oracle = OVM_gasPriceOracle::new(H160(OP_GAS_ORACLE), client.clone());
+                
+                // Get L1 gas used first
```

Comment:
This logic is already encapsulated by the `get_l1_fee` method, so we should just keep that
https://github.com/ethereum-optimism/optimism/blob/3db0492cb4ad37441192f3bc94259cc9df9bb4fc/packages/contracts-bedrock/src/L2/GasPriceOracle.sol#L57

I would suggest investigating if we are using the resulting values correctly, look into the optimism javascript sdk to see how the gas cost is estimated and replicate that here.

Reply:
https://github.com/ethereum-optimism/ecosystem/blob/main/packages/sdk/src/l2-provider.ts

Reply:
If the implementations are identical you can report back.